File: codesignal/happyNumber/happyNumber.ts
function happyNumber(n: number): boolean {
    let memory = new Set()
    let current: number = n

    while (true) {
        console.log(`=== Iteration (current=${current})===`)
        if (current === 1) {
            return true
        } else if (memory.has(current)) {
            return false
        }

        const nStr:string = current.toString()
        let nextValue: number = 0
        for (let c of nStr) {
            let cNum: number = parseInt(c)
            nextValue+=cNum*cNum
            console.log(cNum, nextValue)
        }        
        memory.add(current)
        current = nextValue
    }
}

function main () {
  const testCases = [
    {
      name: 'Example',
      n: 19,
      expect: true
    }
  ]
  for (let test of testCases) {
    const got = happyNumber(test.n)
    if (got !== test.expect) {
        console.error(`[${test.name}] expect ${test.expect}, got ${got}`)
    }
  }
}

main()

File: codesignal/happyNumber/happyNumber.go
package happynumber

import (
	"fmt"
	"strconv"
	"strings"
)

func happyNumber(n int) bool {
	type empty struct{}
	memory := make(map[int]empty)
	current := n

	for {
		fmt.Printf("==== Iteration (current:%v)\n", current)
		if current == 1 {
			return true
		} else if _, ok := memory[current]; ok {
			return false
		}
		nStr := strconv.Itoa(current)
		nextValue := 0
		for _, c := range strings.Split(nStr, "") {
			num, _ := strconv.Atoi(c)
			nextValue += num * num
			fmt.Println(num, nextValue)
		}
		memory[current] = empty{}
		current = nextValue
		// break
	}
	// return false
}
